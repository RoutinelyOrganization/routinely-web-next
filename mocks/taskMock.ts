import { Categories } from '@/types/categories';
import type { Task } from '@/types/task';
import { stringToDate } from '@/utils/formats/stringToDate';

const { longDateString: dateNowStr, day, month, year } = stringToDate();

const dateTomorrow = stringToDate(`${year}-${month}-${day + 1}`);
const dateNextMonth = stringToDate(`${year}-${month + 1}-${day + 1}`);

export const tasks: Task[] = [
  {
    type: 'task',
    id: '1',
    name: 'Tarefa 1',
    checked: false,
    category: Categories.Career,
    date: dateNowStr,
    description: 'teste',
    finallyDate: '2025-11-01 10:00',
    weekDays: ['Monday', 'Friday'],
  },
  {
    type: 'habit',
    id: '2',
    name: 'Tarefa 2',
    checked: false,
    category: Categories.Studies,
    date: dateNowStr,
    description: 'teste',
    finallyDate: '2024-11-01 10:00',
    weekDays: ['Monday', 'Friday'],
  },
  {
    type: 'habit',
    id: '3',
    name: 'Tarefa 3',
    checked: false,
    category: Categories.Career,
    date: dateTomorrow.longDateString,
    description: 'teste',
    finallyDate: dateNextMonth.longDateString,
    weekDays: ['Monday', 'Friday'],
  },
  {
    type: 'task',
    id: '4',
    name: 'Tarefa 4',
    checked: false,
    category: Categories.Studies,
    date: dateTomorrow.longDateString,
    description: 'teste',
    finallyDate: dateNextMonth.longDateString,
    weekDays: ['Monday', 'Friday'],
  },
  {
    type: 'task',
    id: '5',
    name: 'Tarefa 5',
    checked: true,
    category: Categories.Studies,
    date: dateNowStr,
    description: 'teste',
    finallyDate: '2024-11-01 10:00',
    weekDays: ['Monday', 'Wednesday'],
  },
];
