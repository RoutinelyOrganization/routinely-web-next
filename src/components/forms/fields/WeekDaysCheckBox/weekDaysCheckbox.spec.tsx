import type { DaysOfWeek } from '@/types/weekDays';
import { tasks } from '@mocks/taskMock';
import { fireEvent, render, screen } from '@testing-library/react';
import WeekDaysChekbox from '.';

describe('<WeekDaysChekbox/>', () => {
  it('should render', () => {
    render(<WeekDaysChekbox />);

    // busca todos os checkboxes de dias da semana e dispara os eventos de click
    const weekDaysContainer = screen.getByTestId('weekDaysCheckBox');
    const checkboxes = weekDaysContainer.querySelectorAll('input[type="checkbox"]');

    // verifica se os checkboxes foram alterados
    expect(checkboxes[0]).not.toBeChecked();
    expect(checkboxes[1]).not.toBeChecked();
    expect(checkboxes[2]).not.toBeChecked();
    expect(checkboxes[3]).not.toBeChecked();
    expect(checkboxes[4]).not.toBeChecked();
    expect(checkboxes[5]).not.toBeChecked();
    expect(checkboxes[6]).not.toBeChecked();
  });

  it('should render with props', () => {
    render(<WeekDaysChekbox weekDays={tasks[0].weekDays} />);

    const weekDaysContainer = screen.getByTestId('weekDaysCheckBox');
    const checkboxes = weekDaysContainer.querySelectorAll('input[type="checkbox"]');

    for (const checkbox of checkboxes) {
      if (tasks[0].weekDays.includes(checkbox.getAttribute('name') as DaysOfWeek)) {
        expect(checkbox).toBeChecked();
      } else {
        expect(checkbox).not.toBeChecked();
      }
    }
  });

  it('should update the state when the checkbox is clicked', () => {
    render(<WeekDaysChekbox />);

    const weekDaysContainer = screen.getByTestId('weekDaysCheckBox');
    const checkboxes = weekDaysContainer.querySelectorAll('input[type="checkbox"]');

    fireEvent.click(checkboxes[0]);
    fireEvent.click(checkboxes[3]);
    fireEvent.click(checkboxes[5]);

    expect(checkboxes[0]).toBeChecked();
    expect(checkboxes[1]).not.toBeChecked();
    expect(checkboxes[2]).not.toBeChecked();
    expect(checkboxes[3]).toBeChecked();
    expect(checkboxes[4]).not.toBeChecked();
    expect(checkboxes[5]).toBeChecked();
    expect(checkboxes[6]).not.toBeChecked();

    fireEvent.click(checkboxes[0]);
    fireEvent.click(checkboxes[3]);

    expect(checkboxes[0]).not.toBeChecked();
    expect(checkboxes[3]).not.toBeChecked();
  });

  it('should return correct values', () => {
    const setWeekDays = jest.fn();
    render(<WeekDaysChekbox setWeekDays={setWeekDays} />);
    const weekDaysContainer = screen.getByTestId('weekDaysCheckBox');
    const checkboxes = weekDaysContainer.querySelectorAll('input[type="checkbox"]');

    fireEvent.click(checkboxes[0]);
    fireEvent.click(checkboxes[3]);
    fireEvent.click(checkboxes[5]);

    expect(setWeekDays).toHaveBeenCalledTimes(3);
    expect(setWeekDays).toHaveBeenCalledWith(['Sunday', 'Wednesday', 'Friday']);
  });

  it('should macth snapshot', () => {
    const { container } = render(<WeekDaysChekbox />);

    expect(container).toMatchSnapshot();
  });
});
