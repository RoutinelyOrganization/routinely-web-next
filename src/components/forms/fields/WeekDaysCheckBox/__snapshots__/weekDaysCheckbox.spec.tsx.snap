// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeekDaysChekbox/> should macth snapshot 1`] = `
.c2 >input {
  display: none;
}

.c2 >span {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 25px;
  height: 25px;
  border-radius: 50%;
  border: 3px solid #5C59BB;
  cursor: pointer;
}

.c2 >span >img {
  display: none;
}

.c2 >input:checked~span >img {
  display: block;
}

.c0 {
  margin: 8px 0;
  display: flex;
  gap: 8px;
}

.c0 .c1 >span {
  border: 1px solid #5C59BB;
  text-align: center;
  cursor: pointer;
  background-color: #F6F7F8;
}

.c0 .c1 >input:checked~span {
  background-color: #5C59BB;
  color: #F6F7F8;
}

<div>
  <div
    class="c0"
    data-testid="weekDaysCheckBox"
  >
    <label
      class="c1 c2"
      for="checkbox-Sunday"
    >
      <input
        data-testid="checkbox"
        id="checkbox-Sunday"
        name="Sunday"
        role="textbox"
        type="checkbox"
        value="D"
      />
      <span
        aria-label="checkbox"
      >
        D
      </span>
    </label>
    <label
      class="c1 c2"
      for="checkbox-Monday"
    >
      <input
        data-testid="checkbox"
        id="checkbox-Monday"
        name="Monday"
        role="textbox"
        type="checkbox"
        value="S"
      />
      <span
        aria-label="checkbox"
      >
        S
      </span>
    </label>
    <label
      class="c1 c2"
      for="checkbox-Tuesday"
    >
      <input
        data-testid="checkbox"
        id="checkbox-Tuesday"
        name="Tuesday"
        role="textbox"
        type="checkbox"
        value="T"
      />
      <span
        aria-label="checkbox"
      >
        T
      </span>
    </label>
    <label
      class="c1 c2"
      for="checkbox-Wednesday"
    >
      <input
        data-testid="checkbox"
        id="checkbox-Wednesday"
        name="Wednesday"
        role="textbox"
        type="checkbox"
        value="Q"
      />
      <span
        aria-label="checkbox"
      >
        Q
      </span>
    </label>
    <label
      class="c1 c2"
      for="checkbox-Thursday"
    >
      <input
        data-testid="checkbox"
        id="checkbox-Thursday"
        name="Thursday"
        role="textbox"
        type="checkbox"
        value="Q"
      />
      <span
        aria-label="checkbox"
      >
        Q
      </span>
    </label>
    <label
      class="c1 c2"
      for="checkbox-Friday"
    >
      <input
        data-testid="checkbox"
        id="checkbox-Friday"
        name="Friday"
        role="textbox"
        type="checkbox"
        value="S"
      />
      <span
        aria-label="checkbox"
      >
        S
      </span>
    </label>
    <label
      class="c1 c2"
      for="checkbox-Saturday"
    >
      <input
        data-testid="checkbox"
        id="checkbox-Saturday"
        name="Saturday"
        role="textbox"
        type="checkbox"
        value="S"
      />
      <span
        aria-label="checkbox"
      >
        S
      </span>
    </label>
  </div>
</div>
`;
